impress-org/givewp
# Hotfix: prevent lots of extra slashes in the formBuilderSettings meta key which slow down the form builder loading

Resolves [GIVE-92]

## Description



This PR prevents extra slashes from being added to the database on each form builder update and also adds a migration to remove extra slashes in forms that already have it saved on DB.

The problem was happening because the `addSlashesRecursive()` method that was added in [this other PR](https://github.com/impress-org/givewp/pull/6938) to escape double quotes in `emailTemplateOptions` (to fix another bug that was happening in migrated forms) was adding extra slashes on DB on each form update.

There is a similar WP core function called `wp_slash()` that says this in the [docs](https://developer.wordpress.org/reference/functions/wp_slash/):

_This should not be used to escape data going directly into an SQL query._

However, before this PR, we were using the data returned by the `addSlashesRecursive()` method in the Donation Forms repository to update the form settings in the database, which was causing problems with the extra slashes.

So, a possible solution would be to edit the `src/DonationForms/Properties/FormSettings.php` file replacing this line:

```
'emailTemplateOptions' => array_map([$this, 'addSlashesRecursive'], $this->emailTemplateOptions),
```

With this one:

```
'emailTemplateOptions' => wp_slash(wp_unslash($this->emailTemplateOptions)),
```

But we can fully remove this line because it's unnecessary and redundant - this was the approach followed on this PR. 

Why?

The `addSlashesRecursive()` method was originally added to solve a problem that was being generated in the `SanitizeDonationFormPreviewRequest` class related to the `wp_kses_post()` function.

However, this problem related to the use of the `wp_kses_post()` function was fixed a month ago while we were porting the PDF Receipts Addons settings to the form builder. 

Here is the reference commit: https://github.com/impress-org/givewp/commit/dd0230fcc9e659957aa52218738a37f7ccd65c8c#diff-17fad574bc74bd58506d3ab524c91e360df7e2b73648367b983040ef6f740f73

The commit above made the `addSlashesRecursive()` method unnecessary and solved (in the source) the problem that this method was solving, but it kept on the codebase generating extra slashes in the database on each form update.

## Affects



The `formBuilderSettings` meta key.

## Visuals



**Before:** 

![image](https://github.com/impress-org/givewp/assets/16308864/ce626e28-0582-484b-99a6-2ecb16b678a7)

**After:**

![image](https://github.com/impress-org/givewp/assets/16308864/d867e033-e6b1-4bfc-ba6d-ef01e38a13e7)


## Testing Instructions



1. Create a v2 form
2. Migrate the form to v3 (transfer the data)
3. Load design mode, change something, and save the form (refresh the page and do it 2 or 3 times)
4. Check the `formBuilderSettings` meta key on DB and make sure there are no extra slashes there

## Pre-review Checklist






[GIVE-92]: https://stellarwp.atlassian.net/browse/GIVE-92?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
